content/solids/components/voxel-painter.ts

```diff
@@ -0,0 +1,462 @@
+        }
+        
+        let currentDistFromCamera = camera.position.length()
+        camera.updateMatrixWorld()
+        v1.set(0., 0., -currentDistFromCamera)
+        camera.localToWorld(v1)
+        camera.position.sub(v1)
+      }
+
+      if (mouseControlMode === "placing")
+      {
+        const intersection = raycaster.intersectObjects(placingHelpers)[0];
+        if (intersection !== undefined) {
+          placingEnd.copy(intersection.point)
+          snapToNearestValidCubeCenterPosition(placingEnd)
+
+          placementVisualizer.position.addVectors(placingEnd, placingStart).multiplyScalar(.5)
+          placementVisualizer.scale.x = 1. + 2. * Math.abs(placementVisualizer.position.x - placingEnd.x)
+          placementVisualizer.scale.y = 1. + 2. * Math.abs(placementVisualizer.position.y - placingEnd.y)
+          placementVisualizer.scale.z = 1. + 2. * Math.abs(placementVisualizer.position.z - placingEnd.z)
+        }
+      }
+      else
+      {
+        if (mouseControlMode === "rotating" || rotateOnly)
+          placementVisualizer.scale.setScalar(.0000001)
+        else {
+          const intersections = raycaster.intersectObjects(objectsOnWhichVoxelsCanBePlaced);
+          if (intersections.length > 0) {
+            setPositionFromVoxelIntersection(placementVisualizer.position, intersections[0])
+            placementVisualizer.scale.setScalar(1.)
+          }
+          else
+            placementVisualizer.scale.setScalar(.0000001)
+        }
+      }
+
+      if (mouseControlMode === "erasing")
+      {
+        raycaster.ray.at(eraser.position.length(), eraser.position)
+        camera.updateMatrixWorld()
+        camera.worldToLocal(eraser.position)
+
+        const intersections = raycaster.intersectObjects(objectsOnWhichVoxelsCanBePlaced);
+        if (intersections.length > 0 ) {
+          for(let i = 0, il = intersections.length; i < il; ++i)
+          {
+            if (intersections[i].object !== floorIntersectionPlane)
+            {
+              let voxel = intersections[i].object
+              scene.remove(voxel)
+              objectsOnWhichVoxelsCanBePlaced.splice(objectsOnWhichVoxelsCanBePlaced.indexOf(voxel), 1)
+              voxels.splice(voxels.indexOf(voxel), 1)
+            }
+          }
+          setPositionFromVoxelIntersection(placementVisualizer.position, intersections[0])
+          placementVisualizer.scale.setScalar(1.)
+        }
+      }
+      else
+        eraser.position.lerp(defaultEraserPosition,.1)
+    })
+
+    function loop() {
+      let clockDelta = clock.getDelta()
+      frameDelta = clockDelta < .1 ? clockDelta : .1 //clamped because debugger pauses create weirdness
+
+      oldRay.copy(raycaster.ray)
+      raycaster.ray.copy(asyncRaycaster.ray)
+
+      for (var i = 0; i < updateFunctions.length; i++)
+        updateFunctions[i]();
+
+      frameCount++;
+
+      requestAnimationFrame(loop);
+      renderer.render(scene, camera);
+    }
+    loop()
```

Comment:
We don't want an infinite loop like this, since it will keep running even when the element is off-screen. Instead, call the update() function directly within the pointermove callback, when something has actually changed.

Reply:
Done